kolibri/core/tasks/job.py

```diff
@@ -156,9 +172,43 @@ def __getstate__(self):
             "args",
             "kwargs",
             "func",
-            "result",
         ]
-        return {key: self.__dict__[key] for key in keys if key in self.__dict__}
+        working_dictionary = {key: self.__dict__[key] for key in keys if key in self.__dict__}
+
+        if self.exception is not None:
+            working_dictionary["exception_type"] = type(self.exception).__name__
+            if self.exception.args[0] and isinstance(self.exception.args[0], str):
+                working_dictionary["exception_message"] = self.exception.args[0]
+
+        if self.state in [State.CANCELING, State.RUNNING]:
+            logger.debug(
+                "A task is currently in action. This is a bad time to save it to the database."
+            )
+        if self.result is not None:
+            logger.info(
+                "Please know that results of tasks are currently not saved with jobs."
+            )
+
+        try:
+            string_result = json.dumps(working_dictionary, cls=DjangoJSONEncoder)
+        except TypeError as e:
+            # One or more of the keys is not JSON-serializable.
+            logger.debug("Job '{}' is not be saved: {}".format(self.job_id, str(e)))
+            string_result = None
+
+        return string_result
+
+    @classmethod
+    def from_json(cls, json_string):
+        working_dictionary = json.loads(json_string)
```

Comment:
This is where we need to be doing more - we are using the DjangoJSONEncoder above to serialize some Python objects that otherwise will not be serialized by `json.dumps` - we should be doing the inverse here and properly deserializing everything.

Systematically extending `DjangoJSONEncoder` and an equivalent Python `JSONDecoder` object to handle these cases would allow this to be handled very systematically, and for most of the logic of this PR to be focused on this serialization and deserialization step - rather than having to reach into so many other parts of the code.